Render user actions separately

xPaw opened the pull request:
This fixes #577, #404.

Needs to be done before merging this:
- [x] Merge #596
- [x] Merge #597 
- [x] Test and fix and username/action colors if they're terrible
- [x] Look into using font awesome icons for actions

dgw commented:
What I can see of the previews from my phone looks good. Perfect or not, this is a big step forward from how it looks now. :+1:

dgw commented:
See #571, though. The user `button` elements were changed to `a` to allow copying the text in Firefox. The new tpl should use `a`s too.

dgw commented:
Never mind, it moved to #574 and still needs review. But the change will probably be merged.

xPaw commented:
I've tweaked the action texts, and it now looks like this:

![yoloswag_ _shout_-_google_chrome_2016-01-17_22-22-33](https://cloud.githubusercontent.com/assets/613331/12379754/db14394c-bd68-11e5-842e-c9701ca74d03.png)

xPaw commented:
Also formatted /me actions and added "Notice" prefix to notices.

![wowtest_ _shout_-_google_chrome_2016-01-17_22-45-14](https://cloud.githubusercontent.com/assets/613331/12379867/16943500-bd6c-11e5-880c-c9fe93b33d17.png)

MaxLeiter commented:
+1 looks awesome

astorije commented on client/css/style.css:
```diff
@@ -603,11 +603,6 @@ button {
-	color: #ccc;
-	display: none;
-	font-style: normal;
-}
```
@xPaw, what does that remove already?

xPaw replied:
As you can see in the `msg` template, `<em class="type">{{type}}</em>` was removed as it's not needed anymore, it was used to display the action.

astorije commented on client/css/style.css:
```diff
@@ -657,16 +651,6 @@ button {
-#chat .quit .type,
-#chat .topic .type {
-	display: inline;
-}
```
Why this change?

xPaw replied:
Same as above.

astorije commented on client/css/style.css:
```diff
@@ -681,19 +665,38 @@ button {
-	content: "";
+#chat .notice,
+#chat .chan .notice .user {
+	color: #0074D9 !important;
```
What happens if `!important` is absent? Can we easily change something to remove its need? Not that I would hold this PR for this, but really we should get rid of these bastards :-)

xPaw replied:
I tried it, but nickname wasn't colored correctly, I think due to inline nick colors.

astorije commented on client/css/style.css:
```diff
@@ -681,19 +665,38 @@ button {
-#chat .action .user:before {
-	content: "* ";
+#chat .action .from:before {
+	content: "★";
```
Just a random thought, what about using Font Awesome (as we do everywhere else) for these icon-like goodies?

xPaw replied:
Not a terrible idea, but what kind of icons would fit for this purpose? Each different action of course have it's own icon (kicks, topic changes, etc).

astorije commented on client/css/style.css:
```diff
@@ -681,19 +665,38 @@ button {
+	color: #0074D9 !important;
 }
+#chat .notice .user:before {
+	content: "Notice: ";
```
Stupid me, but how do you trigger these? I played with a few commands and stayed on channels around, but it didn't show up.

xPaw replied:
Use `/notice`, but shout doesn't display your own sent notices.

dgw replied:
Easiest way is probably to do `/msg NickServ <something>`, as services on most popular networks usually reply in notices.

astorije replied:
Thanks @dgw, I'll try that.
@xPaw, I guess that's why you opened #590 then :-)

astorije replied:
@xPaw, can we find a Font Awesome icon for this, for consistency with the other actions? Maybe the same as `mode`?
Also, can you post a screenshot of this please? :)

astorije commented on client/js/shout.js:
```diff
@@ -183,14 +183,42 @@ $(function() {
+			case "nick": data.msg.formattedAction = "is now known as"; break;
+			case "part": data.msg.formattedAction = "has left the channel"; break;
+			case "quit": data.msg.formattedAction = "has quit"; break;
+			case "topic": data.msg.formattedAction = "has changed the topic to:"; break;
```
From my experience, topics are often long, full of special characters (colons, dashes, ...), like:

<img width="813" alt="screen shot 2016-01-19 at 23 59 06" src="https://cloud.githubusercontent.com/assets/113730/12440492/fbda6e24-bf09-11e5-85bd-bbdb985a0815.png">

What about displaying the new topic on its own line, like the following:

<img width="817" alt="screen shot 2016-01-20 at 00 03 07" src="https://cloud.githubusercontent.com/assets/113730/12440493/fbdb1d4c-bf09-11e5-96a3-2dc235759b4c.png">

Actually, changing topic is often a big deal that needs to be seen from everyone: I would display it differently to catch the eye, like in bold, or italic, ... What do you think?

MaxLeiter replied:
+1; would like the title to have its own like and be easy to see (although maybe a different color instead of bold/italics?

xPaw replied:
Doing styling is not a good idea as a topic on it's own can contain irc bold/italic/colors.

astorije replied:
What about just the new topic on its own line then?

astorije commented on client/views/msg_action.tpl:
```diff
@@ -0,0 +1,11 @@
+		<a href="#" class="user" style="color: #{{stringcolor from}}">{{mode}}{{from}}</a> {{formattedAction}} {{{parse text}}}
+	</span>
+</div>
+{{/each}}
```
What do you think we should do for the generated templates here (I'm talking specifically here, not long term solution for handling assets)? Adding the `grunt`ed files once this is ready to be merged?

astorije commented:
@xPaw, I left you a bunch of minor comments inline. This looks pretty awesome, really!
I hope we can merge that soon :-)

xPaw commented:
I've added displaying kick/part/quit messages, which will fix #404. However for this to work, `slate-irc` needs to be updated to the latest version, so that the message is passed through correctly.

xPaw commented:
@astorije I've added font awesome icons, looks pretty cool.

astorije commented:
> EDIT: Made topic changes green:

Wheren't you against bold/italic/colors in topic line? I'm OK with anything at the moment really.

astorije commented on client/css/style.css:
```diff
@@ -681,19 +665,58 @@ button {
+}
+#chat .kick .from:before {
+    font-family: FontAwesome;
+	content: "\f00d";
```
For `/kick`, I'm not too fond of this icon, usually used for closing something and sometimes for deleting (but less and less, I notice).
I think I'd prefer `ban` (https://fortawesome.github.io/Font-Awesome/icon/ban/). If really you don't like it, the same icon as leave would work as the consequences are the same (the user leaves the channel) although I still prefer `ban` to that.
Alternatives could be https://fortawesome.github.io/Font-Awesome/icon/exclamation-triangle/, https://fortawesome.github.io/Font-Awesome/icon/exclamation-circle/, https://fortawesome.github.io/Font-Awesome/icon/exclamation/ (but these I would prefer to notify a warning within the app) and if really we want to go esoteric, these could do: https://fortawesome.github.io/Font-Awesome/icon/bolt/, https://fortawesome.github.io/Font-Awesome/icon/crosshairs/, https://fortawesome.github.io/Font-Awesome/icon/crosshairs/, https://fortawesome.github.io/Font-Awesome/icon/thumbs-o-down/, https://fortawesome.github.io/Font-Awesome/icon/frown-o/, https://fortawesome.github.io/Font-Awesome/icon/bomb/.
Meh, `ban` still conveys the most meaning, IMO :-)

astorije commented on client/css/style.css:
```diff
@@ -681,19 +665,58 @@ button {
+}
+#chat .mode .from:before {
+    font-family: FontAwesome;
+	content: "\f013";
```
This one is tricky... But we can't use the `cog` here as it's usually used for settings, and we actually use it for that reason in the app.
Possible alternatives:
- https://fortawesome.github.io/Font-Awesome/icon/info-circle/
- https://fortawesome.github.io/Font-Awesome/icon/key/
- https://fortawesome.github.io/Font-Awesome/icon/terminal/
- https://fortawesome.github.io/Font-Awesome/icon/info/

astorije commented on client/css/style.css:
```diff
@@ -681,19 +665,58 @@ button {
-	content: "* ";
+#chat .action .text,
+#chat .action .user {
+	color: #F39C12 !important;
```
Why this?

astorije commented:
These do look pretty cool... I look forward to that! :-)

xPaw commented:
Changed icons as per feedback

![chrome_2016-01-24_12-01-16](https://cloud.githubusercontent.com/assets/613331/12535655/a6843c66-c292-11e5-8151-377c3d9a4ed7.png)

> Wheren't you against bold/italic/colors in topic line? I'm OK with anything at the moment really.

You are right, I've removed green again. And I would argue it's not the best idea to make it a separate line too.

astorije commented:
I don't think this should have been merged already, while the discussion was ongoing...
